rfcs/0003-api-unification.md

```diff
@@ -0,0 +1,277 @@
+
+This is true but for virtual modules. The virtual module covering
+several "fixed" or "physical" modules and behaves like that would be
+one module. But differences between those physical modules on
+different platforms makes such virtual module a moving target and
+unpredictable.
+
+Virtual modules concept is missing crucial part in the design:
+interfaces. The interface should define how module looks like and what
+APIs can be called to it. Interface should move module that is
+_called_ differently on heterogeneous environments to a module that
+_reports_ differently on heterogeneous environments.
+
+
+# Design
+[design]: #detailed-design
+
+*DISCLAIMER*: The SMI is not that classic understanding of typical
+interface one may find in languages like Java. It is also not as same
+as Zope Interface package or Python Abstract Base Classes (ABC).
+
+The SMI should describe the following properties of the module:
+
+- Functions
+- Signatures
+- Lowest common denominator of the function output format (or minimum
+  required default output structure)
+
+SMI only describes functions of the module and is there to make sure
+that any virtual module is always called exactly the same way, regardless
+what operating system minion is running on.
+
+## Declaration
+[declaration]: #declaration
+
+SMI are declared just as regular Python classes. Salt's "module to
+functions" map is Salt's "interface to methods" map. Therefore `self`
+parameter in the SMI class is not a part of a function signature.
+
+Example of SMI definition for module `pkg`:
+
+```python
+from salt.interfaces import Interface
+
+class PkgInterface(Interface):
+	__modulename__ = 'pkg'
+
+    def list_installed(self, *names, **kwargs):
+        '''
+	    List installed packages.
+	    '''
+		return {}
+
+    def upgrade_available(self, name, **kwargs):
+        '''
+        List available upgrades.
+		'''
+		return {}
+
+    @Interface.supported('weirdlinux', 'beos', 'frogbsd')
+    def salute_fireworks(self, name):
+        '''
+        Launch some fireworks
+        '''
+		return {}
+```
+
+In above incomplete interface example, the list of methods should
+reflect exact names and signatures as in the module, except `self`
+parameter. Rules apply:
+
+- If a method is not in the SMI class, but function is implemented in
+  the module, then such function is marked as "deprecated".
+
+- If a method is in the SMI class but not in the module, then such
+  function is marked as "not implemented".
+
+- If a method has a decorator `@Interface.supported`, only on these
+  platforms unimplemented method will be reported as "not
+  implemented", otherwise "not supported".
+
+
+## Usage
+[usage]: #usage
+
+Once SMI class defined, the usage should be very simple:
+
+```python
+from salt.interfaces.pkg_module import PkgInterface
+
+__virtualname__ = PkgInterface(__name__)()
+
+```
+
+The code above does the following:
+
+- Ensures that the `__virtualname__` is properly set according to the
+  interface.
```

Comment:
How? It looks like we're setting the `__virtualname__` to a class instance. Is the idea to have the loader look for `module.__virtualname__.__modulename__`, and fall back to `module.__virtualname__` if an `AttributeError` is raised?

Reply:
> How? It looks like we're setting the `__virtualname__` to a class instance. 

Ah, no-no. Not the instance. This is doing the following:

```python
__virtualname__ = PkgInterface(__name__)()
```

The `__call__` of the PkgInterface will return just a string. The whole thing here is to:

- `PkgInstance` has already whatever name is inside, so `__call__` will just set it.
- But this call function will also introspect the current module and will readjust everything.

Think of it as just as this:

```python
auto_check_this_module(__name__)
__virtualname__ = 'humasn_writse_tpyos'
```

But I propose to use smart Class instance that also can "disappear" after its init, since it is no longer needed anyway. All is needed is to be a class as an interface carrier, get the module instance to itself, "pipe through" its checks, inject where needed stubs and return a proper name for the virtual module.

In fact, it can be even more simpler:

```python
PkgInterface(__name__)
```

Then it will take current module and set `__virtualname__` by itself. But this is maybe too much magic...